Bluetooth: Mesh: Add h_export support

PavelVPV opened the pull request:
In the current approach the mesh settings will be stored after
the configured timeout. However, there are some cases that require the
settings to be stored immediately, e.g. when power failure has been
detected.

This change allows storing the mesh settings by settings_save() call
by implementing h_export settings handler.

It also adds ability to switch off periodic saving of the settings to
rely only on settings_save() call.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV committed:
Bluetooth: Mesh: Add h_export support

In the current approach the mesh settings will be stored after
the configured timeout. However, there are some cases that require the
settings to be stored immediately, e.g. when power failure has been
detected.

This change allows storing the mesh settings by settings_save() call
by implementing h_export settings handler.

It also adds ability to switch off periodic saving of the settings to
rely only on settings_save() call.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV commented:
This is another attempt to solve issue described in #33257.

PavelVPV commented on subsys/bluetooth/mesh/settings.c:
```diff
@@ -183,7 +130,86 @@ static void store_pending(struct k_work *work)
+{
+	BT_DBG("");
+
+	/* Use settings_save_one because export_func is actually settings_save_one */
```
@nvlsianpu @Laczen @jhedberg @trond-snekvik @alxelax 

We discussed this with @nvlsianpu offline, but I would like you guys to get your opinion on this.

Basically, `export_func` is `settings_save_one`, but implementing it this way may create a potential issue in future in case the settings subsystem is rewritten.

The more proper way of doing that would be to pass a pointer of either `export_func` or `settings_save_one` into `store_pending`. However, it will require extending all the underlying functions...

Laczen replied:
@PavelVPV, I don't quite get what you need the opinion on. I can give you some background on the export functionality: the idea was to provide a way to report the settings, this could be to store them to flash, but it could also be to report them to a user interface or ... 

Doing a settings_save_one is only one of the possibilities.

As it is unknown if this export is save or not (it could be a dump to screen) it is up to the designer to allow or disallow the export functionality by either specifying resp. "nulling" the export function. Most mesh functions do not have a export functionality because it is considered unsafe to share the information.

I don't know if this answers your question, but I would consider it unwise for all mesh settings variables to provide a export function.

PavelVPV replied:
@Laczen, thanks for your reply!

When you are talking about export functionality, do you mean implementation of `h_export` or usage of `export_func`?

I didn't get about the unwise part, because settings subsystem documentation clearly defines what `h_export` is used for and what it uses as `export_func`:
```
A call to settings_save() uses an h_export implementation to store different data in one operation using settings_save_one().
```
Also, do you mean that using `settings_save_one` is not safe either? That sounds like using settings subsystem is unsafe for mesh at all.

What I am trying to emphasise (and asking an opinion for) is that the current design (providing `export_func` in `h_export`) makes some problems in the implementation, which I would like to avoid. On the other hande, I may wrongly understand settings subsystem API (the purpose of `h_export` handler), but then it should be clearly written in the documentation.

alxelax replied:
@PavelVPV, generally, this approach will work but it has a kind of logical inconsistency.
Looks like `h_export` functionality has been designed really to export data to wherever. If it is used for immediate storing everything for the power down case it has to clean up invalidated entries as well. Otherwise, after power-up, they will be restored.  That means it not only stores settings but cleans them either.    
Furthermore, it is unsafe to rely on an external function pointer for storing data, including security credentials. Sounds like a potential security vulnerability.

Laczen replied:
@PavelVPV, as shown in the documentation the `settings_save()` uses an `h_export` implementation to store different data in one operation. This however only works when a `h_export` routine is defined for each `settings_handler`, if there is no `h_export` nothing will be saved.

There used to be a `settings_export()` that you could use to do whatever you wanted with the name and value (e.g. dump to screen). This seems to have been removed along the way.

At the moment the export functionality seems only used for settings_save(). So you use `h_export()` for this, but it might not be easy, each settings_handler is used to retrieve multiple stored items, the settings_handler `h_set` is responsible for decoding the name into the appropiate set routine. For the settings_save() you need to do the opposite and create multiple h_export routines for whatever needs storing.

If your goal is only to ensure that the latest value are written to flash it would be easier to find a way to set the next storage `time` to `now` (and reschedule), or to define a flush to disk routine and call that either directly or based on some other trigger.

PavelVPV replied:
@Laczen, thanks for the clarification, that surprises me. Seems like the only way to achieve the goal is to do it the way you described in your comment.

Laczen replied:
@PavelVPV, I just went over your code again (should have done this more thoroughly before) and your proposed change might break things. The idea behind `settings_save()` is to allow the user to save all settings. Your change only saves very limited amount of data.

I think it would be acceptable to define `settings_save()` as a routine that would allow to save any unsaved data to flash using the `h_export` routine.

PavelVPV replied:
@alxelax I'm not sure I got you. `settings_save_one` or `settings_delete` will be called if any of `bt_mesh_settings_flag` flag is set, which in turns will store or delete pending data.

PavelVPV replied:
@Laczen I think it will save any mesh data for which `bt_mesh_settings_flag` flag is set.

Laczen replied:
> @Laczen I think it will save any mesh data for which `bt_mesh_settings_flag` flag is set.

No need, see my comment above.

Laczen replied:
> @Laczen What if we add here run-time evaluation whether export_func is static_save_on?

@nvlsianpu, this is not needed. Even if in future the use of a export is extended again to provide a dump_to_screen, there is another possibility for the subsystem to protect against bad usage: provide a `h_export()` but do not use the provided export_function. This is exactly what @PavelVPV is doing, so I see no security problem with this.

PavelVPV replied:
@Laczen so should this be reflected in the settings subsystem documentation?

Laczen replied:
@PavelVPV, maybe, but the settings system has a lot of possibilities that are not explicitely described, this is a little difficult at times, but it is also extremely powerful (as shown by yourself). Are we still following the documentation: yes, is there a mentioning that it is obligatory to use the export_routine: no. I would document it in the mesh settings not in the settings, it is just clever usage of the settings module.

Laczen commented:
See my last remark.

Laczen commented:
@nvlsianpu, maybe it is time to (re)introduce settings_export(export_routine) and settings_export_subtree(subtree, export_routine). These could then be called by settings_save() and settings_save_subtree() and could also be used to allow other export methods. The subsystem that uses the export methods has sufficient ways to protect against bad usage (by not providing a h_export, or by not using the provided export routine in h_export).

nvlsianpu commented:
Looks like yes, settings_export_subtree(subtree, export_routine) Is missing, an export function also might be re-introduced (need to consider it more carefully).

PavelVPV committed:
Bluetooth: Mesh: Notify models about storing settings

There are some cases that require the settings to be stored immediately,
e.g. when power failure has been detected. But there is no way to notify
the models that they need to store their user data immediately.

After introducing support for settings_save() in the stack, this can now
be supported at the model level.

Signed-off-by: Pavel Vasilyev <pavel.vasilyev@nordicsemi.no>

PavelVPV commented:
Added model handler to request the settings store at the model level.

jhedberg commented on subsys/bluetooth/mesh/net.c:
```diff
@@ -57,7 +57,7 @@
  * compile the code.
  */
-#ifndef CONFIG_BT_SETTINGS
+#if !defined(CONFIG_BT_SETTINGS) || !defined(CONFIG_BT_MESH_CONTINUOUS_STORE)
```
Doesn't `!defined(CONFIG_BT_MESH_CONTINUOUS_STORE)` cover both, since the option depends on `BT_SETTINGS`?

jhedberg commented on subsys/bluetooth/mesh/settings.c:
```diff
@@ -29,6 +29,11 @@
 #include "settings.h"
 #include "cfg.h"
 
+#ifndef CONFIG_BT_MESH_CONTINUOUS_STORE
+#define CONFIG_BT_MESH_STORE_TIMEOUT 0
+#define CONFIG_BT_MESH_RPL_STORE_TIMEOUT 0
+#endif
```
I see this convention has already crept into some parts of the tree, but I don't think it's a good idea to go synthesizing defines that look like they originate from Kconfig even though they don't actually originate from Kconfig. This can lead to some debugging frustration when you look through the Kconfig definition and can't make sense of why some macro has a value that it has. Instead, I think it would make more sense to use new names in this case, so something like:
```
#ifdef CONFIG_BT_MESH_CONTINUOUS_STORE
#define MESH_STORE_TIMEOUT CONFIG_BT_MESH_STORE_TIMEOUT
#define MESH_RPL_STORE_TIMEOUT CONFIG_BT_MESH_STORE_TIMEOUT
#else
#define MESH_STORE_TIMEOUT 0
#define MESH_RPL_STORE_TIMEOUT 0
#endif
```

jhedberg replied:
Of course the even better solution is to try to ensure that the Kconfig defines aren't attempted to be used in places where they're not relevant. Maybe that's where you're missing #ifdefs?

PavelVPV replied:
Thanks! I would just go with the first option as it allows compiles the code. The parts of the code that shouldn't be included are wrapped into `IS_ENABLED`.

jhedberg commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -660,7 +660,15 @@ config BT_MESH_MODEL_EXTENSIONS
 	  layer to know about Mesh model relationships.
 
-if BT_SETTINGS
+menuconfig BT_MESH_CONTINUOUS_STORE
```
Wouldn't `AUTOMATIC_STORE` be more descriptive, since if you don't set this it means you have to manually request storing?

jhedberg commented on subsys/bluetooth/mesh/Kconfig:
```diff
@@ -660,7 +660,15 @@ config BT_MESH_MODEL_EXTENSIONS
 
-if BT_SETTINGS
+menuconfig BT_MESH_CONTINUOUS_STORE
+	bool "Store mesh settings continuously"
```
Same here. The current implementation isn't really about continuity but ensuring that when something changes that it gets automatically stored.

trond-snekvik commented on subsys/bluetooth/mesh/settings.c:
```diff
@@ -90,42 +74,8 @@ SETTINGS_STATIC_HANDLER_DEFINE(bt_mesh, "bt/mesh", NULL, NULL, mesh_commit,
-}
-
-static void store_pending(struct k_work *work)
+static void store_pending(void)
```
In lieu of a `settings_store_subtree` API, perhaps it makes sense to make this a public API in mesh/main.h?
Calling `settings_save()` pulls in all modules with settings storage, which could make it impossible to determine a worst case storage time.

PavelVPV replied:
These are good questions! If we only need to store critical parameters of the stack then I think `settings_save()` is a wrong way and we need to make `store_pending` public and forget about `settings_save()`. Because I think if `settings_save()` supported by the stack, it should propagate this to the models, otherwise there is no other way to handle this at the models level.

However, what would be impact of not saving the models' data if power loss detected? And what if there are some other settings that need to be stored?

Laczen replied:
@PavelVPV, @trond-snekvik, please forget about the powerloss case, it will not be possible to sustain the power long enough to be stored. For nvs with n sectors the worst case would be to do n sector **erases** before doing one save. This method would be good for planned shutdown, but a direct call to `store_pending` might be a better solution.

trond-snekvik commented on subsys/bluetooth/mesh/settings.c:
```diff
@@ -183,7 +133,92 @@ static void store_pending(struct k_work *work)
+	 * In this case, it can be unsafe to use export_func. Since it is not obligatory
+	 * to use export_func when implementing h_export, use settings_save_one directly.
+	 */
+	store_pending();
```
Would it make sense to check whether `export_func` actually is `settings_save_one` before calling store_pending? If settings_export is reintroduced, this will currently store all data whenever someone tries to dump it to a log or something

PavelVPV replied:
Yes, I think, that would make sense.

trond-snekvik reviewed (commented):
Pretty neat, the overall approach looks good to me.
